contracts/sol/MACI.sol

```diff
@@ -191,20 +239,138 @@ contract MACI is Ownable, DomainObjs {
         emit PublishMessage(_message, _encPubKey);
     }
 
-    //// Submits proof for updating state tree
-    //function verifyUpdateStateTreeProof(
-      //uint[2] memory,
-      //uint[2][2] memory,
-      //uint[2] memory,
-      //uint[28] memory 
-    //) public pure returns (bool) {
-      ////return batchUstVerifier.verifyProof(a, b, c, input);
-      //// TODO: submit a batch of messages
-      //return true;
-    //}
+    function unpackProof(
+        uint256[8] memory _proof
+    ) public pure returns (
+        uint256[2] memory,
+        uint256[2][2] memory,
+        uint256[2] memory
+    ) {
+
+        return (
+            [_proof[0], _proof[1]],
+            [
+                [_proof[2], _proof[3]],
+                [_proof[4], _proof[5]]
+            ],
+            [_proof[6], _proof[7]]
+        );
+    }
+
+    /*
+     * A helper function to create the publicSignals array from meaningful
+     * parameters.
+     * @param _newStateRoot The new state root after all messages are processed
+     * @param _stateTreeRoots The intermediate state roots
+     * @param _ecdhPubKeys The public key used to generated the ECDH shared key
+     *                     to decrypt the message
+     */
+    function genBatchUstPublicSignals(
+        uint256 _newStateRoot,
+        uint256[] memory _stateTreeRoots,
+        PubKey[] memory _ecdhPubKeys
+    ) public view returns (uint256[19] memory) {
+        uint256[19] memory publicSignals;
+        publicSignals[0] = _newStateRoot;
+        publicSignals[1] = coordinatorPubKey.x;
+        publicSignals[2] = coordinatorPubKey.y;
+        publicSignals[3] = voteOptionsMaxLeafIndex;
```

Comment:
This does not need to be public input you can just set it in the snark circuit at compile time.

Reply:
Let's say the vote option tree has 4 levels (max 16 leaves). However, there are only 10 options to vote for (e.g. if you want to vote for what to buy at a restaurant, the menu may only have 10 items).

**Scenario A**

Alice votes for option #0 (which is valid) and Bob votes for option 15 (which is also valid). Both votes are counted.

At the end of the day, the coordinator tallies the votes and ignores votes for options 10-15, so only Alice's vote has an effect.

**Scenario B**

Alice votes for option #0 (which is valid) and Bob votes for option 15 (which is invalid).

Alice's vote is counted, but Bob's vote is a no-op.

At the end of the day, the result is the same as Scenario A.

However, Scenario B will allow us to spend less gas, since we can reduce the number of public inputs by 1. As such, in the grand scheme of things, I think we should go with Scenario B.

Reply:
Well if a user wants to charge you gas they can put noop votes on chain that you are required to process. So don't agree with argument about reducing gas costs.  

NOTE : I think if we want to reduce gas we should make an optimisitic version of MACI.

Reply:
Correction: Scenario A (not B) allows us to spend less gas.

But after some discussion, we've decided to go with B.